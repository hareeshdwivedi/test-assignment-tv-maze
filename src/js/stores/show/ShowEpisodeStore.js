import { observable, action, extendObservable } from 'mobx';
import Store from 'stores/Store';
import endpoints from 'transport/endpoints';

/**
 * The show episode store
 */
class ShowEpisodeStore extends Store {

    /** The show's episodes */
    @observable
    showEpisodes = {};

    /**
     * Get a show's episodes
     * @returns {Number} show's episodes
     */
    get ShowEpisodes() {
        return this.showEpisodes;
    }

    /**
     * Set a show's episode
     * @param {Object} showEpisode - show's episode to be set
     * @param {Array} showEpisodeParams - show's episode params
     */
    @action
    setShowEpisode = (showEpisode, showEpisodeParams) => {
        this.showEpisodes[showEpisodeParams.join('_')] = showEpisode;
    };

    /**
     * Fetch a show's episode
     * @param {Number} showId - a show identifier
     * @returns {Promise} a call api promise
     */
    fetchShowEpisode(showId, seasonId, episodeId) {
        const showEpisodeParams = [showId, seasonId, episodeId];
        return this.Transport.callApi(endpoints.GET_SHOW_EPISODE, null, showEpisodeParams)
            .then(showEpisode => this.setShowEpisode(showEpisode, showEpisodeParams));
    }
}

export default ShowEpisodeStore;