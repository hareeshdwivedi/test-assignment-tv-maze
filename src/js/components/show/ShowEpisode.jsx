import React, { Component } from 'react';
import { observer } from 'mobx-react';
import PropTypes from 'prop-types';

import ShowEpisodeStore from 'stores/show/ShowEpisodeStore';

import * as styles from 'css/show-episode.less';

/**
 * The component shows details of a show's episode
 */
@observer
class ShowEpisode extends Component {

    /** The component properties */
    static propTypes = {
        store: PropTypes.instanceOf(ShowEpisodeStore)
    };

    /**
     * @see Component#componentWillMount()
     */
    componentWillMount() {
        const { params: { showId, seasonId, episodeId }, store } = this.props;
        store.fetchShowEpisode(showId, seasonId, episodeId);
    }

    /**
     * @see Component#render()
     */
    render() {
        const { store: { ShowEpisodes }, params: { showId, seasonId, episodeId }} = this.props;
        const showEpisode = ShowEpisodes[`${showId}_${seasonId}_${episodeId}`];
        if (showEpisode) {
            const { name, image, summary} = showEpisode;
            return (
                <div className="rtl-episode">
                    <img src={image.original} />
                    <h1>{name}</h1>
                    <p dangerouslySetInnerHTML={{__html: summary}} />
                </div>
            );
        }
        return null;
    }
}

export default ShowEpisode;
