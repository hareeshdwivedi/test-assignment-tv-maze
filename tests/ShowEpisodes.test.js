import React from 'react';
import renderer from 'react-test-renderer';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import ShowEpisodeStore from 'src/js/stores/show/ShowEpisodeStore';
import ShowEpisode from 'src/js/components/show/ShowEpisode';

describe('Test ShowDetails component', () => {

    configure({adapter: new Adapter()});

    it('check initial markup of the component', () => {
        const mockedStore = new ShowEpisodeStore();
        mockedStore.fetchShowEpisode = jest.fn();
        mockedStore.setShowEpisode(
            {name: 'name1', image: {original: 'url1'}, summary: 'summary'},
            [1, 1, 1]
        );

        const renderedComponent = renderer.create(
            <ShowEpisode params={{showId: 1, seasonId: 1, episodeId: 1}} store={mockedStore}/>
        );
        expect(renderedComponent.toJSON()).toMatchSnapshot();
    });
});