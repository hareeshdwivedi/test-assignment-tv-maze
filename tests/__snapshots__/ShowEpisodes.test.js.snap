// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ShowDetails component check initial markup of the component 1`] = `
<div
  className="rtl-episode"
>
  <img
    src="url1"
  />
  <h1>
    name1
  </h1>
  <p
    dangerouslySetInnerHTML={
      Object {
        "__html": "summary",
      }
    }
  />
</div>
`;
